One student module history per request

jcohen28 opened the pull request:
<!--

🫒🫒
🫒🫒🫒🫒         🫒 Note: the Olive master branch has been created.  Please consider whether your change
    🫒🫒🫒🫒     should also be applied to Olive. If so, make another pull request against the
🫒🫒🫒🫒         open-release/olive.master branch, or ping @nedbat for help or questions.
🫒🫒

🌰🌰🌰🌰🌰🌰     🌰 Note: the Nutmeg release is still supported.
                  Please consider whether your change should be applied to Nutmeg as well.

Please give your pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html

Use this template as a guide. Omit sections that don't apply. You may link to information rather than copy it.
More details about the template are at https://github.com/openedx/open-edx-proposals/pull/180
(link will be updated when that document merges)
-->

NOTE: THIS SOLUTION IS DEPENDENT ON #31189

## Description

When a student submits a problem answer, the state is stored in a StudentModule record containing answer, score, correctness, etc. The record, though, is updated in multiple steps within the single request (first the grade is updated, then the state is updated separately). Each partial save would trigger a separate StudentModuleHistory record to be stored resulting in duplicate and inaccurate historical records.

This solution uses the RequestCache to track within a request thread which StudentModules are updated and a single corresponding StudentModuleHistory id. If multiple update actions occur within the request cycle, then modify the history record that was already generated to ensure that each submission only results in one StudentModuleHistory record.

## Supporting information

This issue and its solution were discussed [here](https://discuss.openedx.org/t/extra-history-record-stored-on-each-problem-submission/8081)

## Testing instructions

Submit an answer for any problem xblock, then confirm that the `courseware_studentmodulehistory` table only contains one new row corresponding to the latest submission.

## Deadline

None

## Other information

None

ziafazal committed:
fix: Fixed broken HiddenContentTransformer when course pacing is self_paced

fix: Fixed broken HiddenContentTransformer when course pacing is self_pace

In case of self paced course we are relying on course end date to determine if a block is visible or not. However, when course blocks API is called with a block_key it breaks since all blocks do not have end field instead only course block has end field. These Changes fetch end date of block from parent or ancestors blocks to make sure we have course end date available while determining block visibility.

ziafazal committed:
fix: Fixed broken HiddenContentTransformer when course pacing is self_paced. see commit 1ebe0e6 too

jcohen28 committed:
fix: Only create one StudentModuleHistory record per request

A StudentModule instance may get saved multiple times in one request cycle, but should still only trigger one StudentModuleHistory entry

Custom RequestCache namespace + logging

remove logging

cleanup

varshamenon4 committed:
feat: show exam info in course outline

UvgenGen committed:
feat!: Remove inheritance-related code from Old Mongo

navinkarkera committed:
fix: preview mode navigation previous button (that uses legacy UI)

Legacy UI: Previous button should use prev_url if it in first
position. In studio UI, it should always use prev_url. To make this
possible we checking length of contents which is always zero for studio
and more than 0 for legacy UI.

openedx-webhooks commented:
Thanks for the pull request, @jcohen28!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

muhammad-ammar committed:
chore: upgrade edx-enterprise to 3.58.0

Add a new endpoint to generate a signed token for plotly analytics.

jansenk committed:
fix: remove psudosessioncourseserializer as we now have access to courseoverviews (#31187)

Commit:
feat: remove constraint on and upgrade openedx-events, edx-event-bus-kafka (#31182)

This will remove the "effort" field from COURSE_CATALOG_INFO_CHANGED event data

Commit:
chore(i18n): update translations

zacharis278 committed:
chore: update edx-proctoring python and js (#31185)

aht007 committed:
chore: install xblock-drag-and-drop from pypi (#31196)

bcitro committed:
chore: bump enterprise version

NIXKnight committed:
chore: update Dockerifle to run apps as a low privileged user (#31179)

nsprenkle committed:
temp: added debug logging to init view (#31204)

Also involved reordering some code into more logical groupings.

aht007 committed:
Install xblock-google-drive from pypi (#31201)

* chore: install xblock-google-drive from pypi

ashultz0 committed:
fix: pick up a due date fix in proctoring 4.13.0

Commit:
chore: geoip2: update maxmind geolite country database

UsamaSadiq committed:
fix: bump pylint version (#31084)

aht007 committed:
chore: install codejail from PyPI (#31207)

Commit:
feat: add discussion reported and unreported events

aht007 committed:
Revert "chore: install codejail from PyPI (#31207)" (#31213)

This reverts commit 66e4902e9ed017a1562b16457562362c2c529010.

connorhaugh committed:
fix: studio edit permissions (#257)

Per 12. Make your fix public. 
I am merging this fix.

nsprenkle committed:
feat: add timers / logs for for profiling learner home (#31210)

* feat: added timer func for profiling learner home

Also involved proxying some imported functions to be able to time.

ashultz0 committed:
feat: proctoring to 4.13.1 to pick up two to one year change

alangsto committed:
feat: update docker production log settings (#31217)

nsprenkle committed:
refactor: course resume URL speed enhancements (#31222)

* refactor: remove resume url checks for performance

Existing "get resume urls" loads the course structure to verify the last
completed block actually exists. This is a potentially very expensive
operation, especially for users with many enrollments. This isn't a huge
issue since Learning MFE already has a fallback behavior for this case.

* refactor: remove now-unnecessary serializer check

Old behavior passed an empty string for missing resume URL and
serializer would cast to None. Instead, update simply passes through
None for missing resume url

* style: update dependency ordering

* style: run black on some more files

UvgenGen committed:
test: prepare tests for removing support for children in Old Mongo

Commit:
refactor: remove todo to add tracking for reported content email notification

zacharis278 committed:
test: update exams test to use split modulestore (#31223)

ashultz0 committed:
feat: update lti xblock to refactored version 6.0.0

alex-sheehan-edx committed:
chore: bumping enterprise version to 3.58.3

syedsajjadkazmii committed:
feat: create new waffle flag for learner home recommendations [VAN-1138] (#31225)

alex-sheehan-edx committed:
chore: bumping enterprise version to 3.58.4

Commit:
chore(i18n): update translations (#31231)

Co-authored-by: Jenkins <sre+jenkins@edx.org>
Co-authored-by: Tim McCormack <tmccormack@edx.org>

Commit:
chore(i18n): update translations

nedbat commented:
📣 💥 Heads-up: You must either **rebase onto master** or **merge master into your branch** to keep passing required checks.

We added a new required check, "Tests Successful," that this PR does not yet run.  Rebasing will get it started.

If you have any questions, please reach out to the Architecture team (either [#architecture](https://openedx.slack.com/archives/C024JPE27PV) on Open edX Slack or [#external-architecture](https://edx-internal.slack.com/archives/C024JPE27PV) on edX internal Slack).

xitij2000 committed:
fix: Change approach to processing course topics (#31200)

Course topics are now created by traversing the entire course structure from top to bottom instead of starting at the sequential level and then moving up or down as needed.

This also introduces a lot of debug logs to pontetially find the reason why under some circumstances new units don't get processed and end up without a discussions topic.

Commit:
feat: update event bus dependencies to bring in schema fix (#31237)

Address linter feedback

spacing

linter feedback

nsprenkle committed:
fix: get audit expiration from correct data source (#31238)

Commit:
chore: upgrade packages, specifically edx-braze-client

Commit:
feat: upgrade edx-event-bus-kafka to bring in new error logging and monitoring (#31245)

ashultz0 committed:
fix: update proctoring library for better expiration date sharing

nsprenkle committed:
refactor: learner home cleanup (#31240)

* refactor: remove dev logging

* style: run black

* refactor: move masquerade check into utils

* style: remove unnecessary assignment

* style: update dependency orderings

* refactor: add function tracing for perf testing

* refactor: move grade data fetching out of serializer

This allows us to separately profile the collection of grade data

* fix: add missing requires_context metadata

* refactor: split out serialization for profiling

kiram15 committed:
fix: version update

Commit:
chore: geoip2: update maxmind geolite country database

UsamaSadiq committed:
Install pip and pip tools (#31206)

* fix: install pip & pip-tools in make upgrade
* chore: Updating Python Requirements (#31205)

Co-authored-by: edX requirements bot <49161187+edx-requirements-bot@users.noreply.github.com>

muhammadadeeltajamul committed:
feat: added post filters in learners tab api (#31191)

Co-authored-by: adeel.tajamul <adeel.tajamul@arbisoft.com>

jansenk committed:
fix: remove unused url from backend response (#31244)

Commit:
feat: Ping tCRIL on-call if someone updates the legacy dashboard.

This could be a person using the quickstart and we want to be notified
of that so we can prais them and close the PR.

nsprenkle committed:
feat: implement missing audit access expired check (#31249)

Commit:
fix: Remove non-working teams from the CODEOWNERS file.

The teams stopped working when we moved the repo to the openedx org.
Also, things have been re-organized at 2U so it's unclear if these
ownership mappings are still relevant.  I'm removing them for now and
I've posted on the forums to let people know if they want to add back in
more relevant teams.

Relevant Link: https://discuss.openedx.org/t/edx-platform-codeowners/8649

syedsajjadkazmii committed:
fix: add fix for XSS Vulnerability and Open Redirect [VAN-1133] (#259)

jcohen28 committed:
fix: Account for caching logic in unit tests

openedx-webhooks commented:
@jcohen28 Even though your pull request wasn’t merged, please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/31190&entry.1484655318=https://github.com/jcohen28&entry.752974735=2022-10-20+19:29&entry.1917517419=2022-11-07+13:45&entry.2133058324=No) so we can improve your experience in the future.

<!-- comment:end_survey -->
<!-- data: eyJqaXJhLXByZS1jbG9zZSI6IG51bGx9 -->
